lib/BackbeatProducer.js

```diff
@@ -128,17 +128,28 @@ class BackbeatProducer extends EventEmitter {
                 cb(errors.InternalError);
             });
         }
-        const payload = this._partition === undefined ?
-                  entries.map(item => ({
-                      topic: this._topic,
-                      messages: new KeyedMessage(item.key, item.message),
-                      key: item.key,
-                  })) :
-                  entries.map(item => ({
-                      topic: this._topic,
-                      messages: item.message,
-                      partition: this._partition,
-                  }));
+
+        const payload = entries.map(item => {
+            if (this._partition === undefined && item.partition === undefined) {
+                return {
+                    topic: this._topic,
+                    messages: new KeyedMessage(item.key, item.message),
+                    key: item.key,
+                };
+            } else if (item.partition && this._partition === undefined) {
```

Comment:
You can get rid of the `else`.

Minor style comment: the tests would be simpler and more understandable if you use the following sequence instead:

```
if (this._partition) { return ...}
if (item.partition) { return ...}
return {...}
```